service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/bucket/CopyGcsBucketDefinitionStep.java

```diff
@@ -0,0 +1,158 @@
+            sourceBucket.getManagedBy(),
+            bucketName);
+
+    final ApiGcpGcsBucketCreationParameters destinationCreationParameters =
+        getDestinationCreationParameters(inputParameters, workingMap);
+
+    final List<ControlledResourceIamRole> iamRoles = getIamRoles(sourceBucket.getAccessScope());
+    final ControlledGcsBucketResource clonedBucket =
+        controlledResourceService.createBucket(
+            destinationBucket, destinationCreationParameters, iamRoles, userRequest);
+    final ApiCreatedControlledGcpGcsBucket apiCreatedBucket =
+        new ApiCreatedControlledGcpGcsBucket()
+            .gcpBucket(clonedBucket.toApiResource())
+            .resourceId(destinationBucket.getResourceId());
+    final ApiClonedControlledGcpGcsBucket apiBucketResult =
+        new ApiClonedControlledGcpGcsBucket()
+            .bucket(apiCreatedBucket)
+            .sourceWorkspaceId(sourceBucket.getWorkspaceId())
+            .sourceResourceId(sourceBucket.getResourceId());
+    workingMap.put(JobMapKeys.RESPONSE.getKeyName(), apiBucketResult);
+    return StepResult.getStepResultSuccess();
+  }
+
+  private ApiGcpGcsBucketCreationParameters getDestinationCreationParameters(
+      FlightMap inputParameters, FlightMap workingMap) {
+    final ApiGcpGcsBucketCreationParameters sourceCreationParameters =
+        workingMap.get(
+            ControlledResourceKeys.CREATION_PARAMETERS, ApiGcpGcsBucketCreationParameters.class);
+    final Optional<String> suppliedLocation =
+        Optional.ofNullable(inputParameters.get(ControlledResourceKeys.LOCATION, String.class));
+
+    // Override the location parameter if it was specified
+    if (suppliedLocation.isPresent()) {
+      return new ApiGcpGcsBucketCreationParameters()
+          .defaultStorageClass(sourceCreationParameters.getDefaultStorageClass())
+          .lifecycle(sourceCreationParameters.getLifecycle())
+          .name(sourceCreationParameters.getName())
+          .location(suppliedLocation.get());
+    } else {
+      return sourceCreationParameters;
+    }
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    final ControlledGcsBucketResource clonedBucket =
+        flightContext
+            .getWorkingMap()
+            .get(
+                ControlledResourceKeys.CLONED_RESOURCE_DEFINITION,
+                ControlledGcsBucketResource.class);
+    controlledResourceService.deleteControlledResourceSync(
+        clonedBucket.getWorkspaceId(), clonedBucket.getResourceId(), userRequest);
+    return StepResult.getStepResultSuccess();
+  }
+
+  private String randomBucketName() {
+    return UUID.randomUUID().toString().toLowerCase();
+  }
+
+  /**
+   * Build the list of IAM roles for this user. If the resource was initially shared, we make the
+   * cloned resource shared as well. If it's private, the user making the clone must be the owner
+   * and becomes EDITOR on the new resource.
```

Comment:
EDITOR on private resources allows a user to delete the resource but not read or write, which isn't very useful. Can we copy the user's roles from the old resource?

Reply:
Assigning the cloning user all three roles, as the copy is theirs to do with as they will.